internal/runner/nomad_manager.go

```diff
@@ -190,9 +190,21 @@ func (m *NomadRunnerManager) onAllocationStopped(alloc *nomadApi.Allocation) {
 		return
 	}
 
+	_, isUsed := m.usedRunners.Get(alloc.JobID)
 	m.usedRunners.Delete(alloc.JobID)
+
 	environment, ok := m.environments.Get(environmentID.ToString())
 	if ok {
 		environment.DeleteRunner(alloc.JobID)
+
+		if !isUsed && failed {
+			entry := log.WithField("failed_job", alloc.JobID)
+			if runner, ok := environment.Sample(); ok {
+				environment.AddRunner(runner)
```

Comment:
What is the reasoning behind sampling and re-adding the runner just sampled? If I read the code correctly, you get a clean (unused) runner first, just to re-add it to the list of available unused runners the line after. Is this just to trigger the re-creation of a new runner?

I assume it is probably related to your statement:
> We restart only every failed, not-used allocation of environments that has idle runners (/ at least one successful runner). This catches allocation failures due to the job specification (such as inaccessible Docker image).

In any way, if we decide to keep this system, we should probably add a suitable test case here (since this will mechanism will be required in production).